### ISSUE TITLE
maven-metadata.xml SHA256 and SHA512 checksums prevent publishing to Nexus

### ISSUE THREAD

**Issue by 120011676** [gradle/gradle#11308]
### Expected Behavior
 publish Maven Repository success

### Current Behavior

<img width="1680" alt="Screen Shot 2019-11-11 at 18 51 30" src="https://user-images.githubusercontent.com/1992933/68582114-f3da0f00-04b4-11ea-8a07-c64e54530b53.png">

···
failureMessage | Artifact updating: Repository ='releases:Releases' does not allow updating artifact='/com/github/120011676/vine/maven-metadata.xml.sha512'
-- | --
failureMessage | Artifact updating: Repository ='releases:Releases' does not allow updating artifact='/com/github/120011676/vine/maven-metadata.xml.sha256'
···

### Context

 publish Maven Repository

### Steps to Reproduce 

https://github.com/120011676/vine

### Your Environment

Build scan URL:https://github.com/120011676/vine

**Comment by melix** [comment]
Does Gradle fail the upload task or is it emitted as a warning? Normally if the repository doesn't support sha512 we continue.

**Comment by 120011676** [comment]
@melix Maven central warehouse used,5.6.3 did not appear, how can I succeed?

**Comment by melix** [comment]
Is it just me or you are using a GitHub package registry here? There's a known issue with GitHub. We're working with them so that they fix it.

**Comment by melix** [comment]
I'm going to ping Sonatype and link to this issue, there may be something wrong with the validation of `maven-metadata.xml`.

**Comment by 120011676** [comment]
@melix Fabulous,Yes, I have a comment

**Comment by 120011676** [comment]
@melix It's like I found out why

maven md5: 91031dce42877df3835d898ac4f80da7
gradle md5: 2d7df10d9d5866df3be7fdc92f7cedc7

maven sha1: 63acee5ae8665b78c3159e79abb0b04f72f1be56
gradle sha1: 95acfd83889e54415725a3964bf7272b40d5d4ae

maven sha256: 14552cce3c16c3c29ecd8d6ca17b8b38529277add5930589e0d47f20a63af213
gradle sha256: 822b4e3bfdbac894dc7fc75209dd749c5795657de06f4e363ef9497e7bcfad08

maven sha512: 9d44c0e17dc3518cd6057d771b99dec7bf5029fefc36fd2287b54cd1fbfe20ade99b23983da3aa3f531fda98d13b2c0bd60af65d7161cc96408406d7347708d3
gradle sha512: 36f00e22b09b6f503be056d53ac4c3210a8f6b67c353ba18acd64b9740885dc8606304909eb6f8b4dac40039ce19ebb20a657e3487a3f291f05d86ab51383a06

https://repo.maven.apache.org/maven2/com/github/120011676/vine/

**Comment by vlsi** [comment]
Does that mean Gradle 6.0.1 could have a feature-flag to avoid publishing sha256 / sha512 for artifacts and metadata?

**Comment by melix** [comment]
Sonatype is planning a fix. If it's too late to ship, we'll have to consider a flag, but we'd like to avoid this situation.

**Comment by vlsi** [comment]
> but we'd like to avoid this situation.

Exactly

> If it's too late to ship, we'll have to consider a flag

A bit of a story is "Nexus vs Gradle 6" issue did hit us when trying to release Apache JMeter yesterday.

Apache Software Foundation is still using Nexus 2.x, and it would take time to make ASF sha512-compatible. It would require either Nexus 2->3 upgrade, or back-porting of the fix or migrate to Artifactory or whatever.

I beg your pardon, but I'm inclined to request a feature flag (even though I recognize the year is 2019) as the upgrade of ASF infra might take a while (there was [an attempt in 2016-2017](https://issues.apache.org/jira/browse/INFRA-12886), and it was just closed). In the meantime, I've [asked re ASF Nexus upgrade](https://issues.apache.org/jira/browse/INFRA-14923?focusedCommentId=16973071&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16973071)

**Comment by melix** [comment]
Ok so let's prioritize a magic flag for 6.0.1.

**Comment by 120011676** [comment]
Like, encourage upgrading

**Comment by melix** [comment]
Before I merge the workaround, I'd like to double check that the Gradle build indeed did *not* fail, and that's only in Nexus that you're seeing the problem.

**Comment by 120011676** [comment]
@melix One other bug found,I don't know

**Comment by jjohannes** [comment]
@120011676 we have included a flag for a `6.0.1` patch release. Here is a nightly containing the chage:
https://services.gradle.org/distributions-snapshots/gradle-6.0.1-20191115004001+0000-bin.zip
It would be very much appreciated if you can acknowledge that it works for you.

To disable publication of both SHA-256 and SHA-512 checksums, either:
- add `-Dorg.gradle.internal.publish.checksums.insecure` to the CLI or
- add `org.gradle.internal.publish.checksums.insecure=true` to your
`gradle.properties` file

**Comment by vlsi** [comment]
@jjohannes , are you sure that snapshot includes the fix?

I've tried gradle-6.0.1-20191115004001+0000-all.zip , and it still publishes sha256 and sha512 files.
Then I opened `/.gradle/wrapper/dists/gradle-6.0.1-20191115004001+0000-all/8vxyfuyndjkwgc528crr43wp/gradle-6.0.1-20191115004001+0000`, and I don't see the fix there:

```
$ grep publishChecksums -A 8 ExternalResourceResolver.java
        publishChecksums(destination, src);
    }

    private void publishChecksums(ExternalResourceName destination, File content) {
        publishChecksum(destination, content, "sha1", 40);

        publishPossiblyUnsupportedChecksum(destination, content, "sha-256", 64);
        publishPossiblyUnsupportedChecksum(destination, content, "sha-512", 128);
    }

    private void publishPossiblyUnsupportedChecksum(ExternalResourceName destination, File content, String algorithm, int length) {
        try {
```

Is it what you would expect?

Just in case, I've tried to specify all the possible options at the same time (`-D`, `-P`, and the option in `gradle.properties`), and Gradle still publishes `sha256` and `sha512`.

If that matters, I tested both `-all` and `-bin` snapshots, and they both publish sha256/512 :-/

**Comment by ljacomet** [comment]
Sorry, two things at play here:
* First a copy/paste mistake in the nightly version. Please use `6.0.1-20191115103811+0000` as the version, or `https\://services.gradle.org/distributions-snapshots/gradle-6.0.1-20191115103811+0000-bin.zip` for the `gradle-wrapper.properties`
* The instructions need to be corrected:
  * Adding `-Dorg.gradle.internal.publish.checksums.insecure=true` on the CLI
  * Adding `systemProp.org.gradle.internal.publish.checksums.insecure=true` to `gradle.properties`

**Comment by vlsi** [comment]
Thanks for the clarification.

I've tried `6.0.1-20191115103811+0000`, and I confirm `systemProp.org.gradle.internal.publish.checksums.insecure=true` works: it does not produce sha256/512 for  artifacts / maven-metadata.xml / module files.

PS I verify with a Nexus stub ( https://github.com/vlsi/asflike-release-environment ), so the mileage may vary.

**Comment by phreakadelle** [comment]
Thanks for providing the switch. 

We were forced to use the `-Dorg.gradle.internal.publish.checksums.insecure=true` parameter as we are using Gradle 6.5.1 and Nexus 2.14.15-01

**Comment by gavenkoa** [comment]
I hit the problem with Nexus OSS 3.28.1 and `systemProp.org.gradle.internal.publish.checksums.insecure=true` in `gradle.properties` is a solution.

According to https://issues.sonatype.org/browse/NEXUS-21802 fix is in v2.14.18 but not in v3.x:

> MSDE Team yes, the support was added to NXRM2 but hasn't yet been done in NXRM3, NEXUS-23603 is the ticket in question, please do add a vote on there

3.x still has the issue, see: https://issues.sonatype.org/browse/NEXUS-23603

**Comment by alexeyOnGitHub** [comment]
does this "-Dorg.gradle.internal.publish.checksums.insecure=true" flag turn off all sha publishing or just for 256 and 512?
